### ISSUE TITLE
OpenAPI Schema Generation is Still Broken for Nested Discriminated Unions

### ISSUE THREAD

**Issue by LukeHartmanTmlt** [pydantic/pydantic#7491]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

I found a related issue, https://github.com/pydantic/pydantic/issues/5436, which ran into this for V1, but this issue is also present in V2.

I was able to use the same code, with very minor modifications to replicate this behavior: I changed .schema() to .model_json_schema(), and tweaked the comment to show the new output

### Example Code

```Python
from enum import Enum
from typing import Literal, Union

from typing_extensions import Annotated

from pydantic import BaseModel, Field, ValidationError

class PetType(str, Enum):
    CAT = "cat"
    DOG = "dog"

class Color(str, Enum):
    WHITE = "white"
    BLACK = "black"

class BasePet(BaseModel):
    pet_type: PetType

class BaseCat(BasePet):
    pet_type: Literal[PetType.CAT]
    color: Color


class BlackCat(BaseCat):
    color: Literal[Color.BLACK]
    black_name: str


class WhiteCat(BaseCat):
    color: Literal[Color.WHITE]
    white_name: str


Cat = Annotated[Union[BlackCat, WhiteCat], Field(discriminator='color')]

class Dog(BasePet):
    pet_type: Literal[PetType.DOG]
    barks: float

Pet = Annotated[Union[Cat, Dog], Field(discriminator='pet_type')]

class Model(BaseModel):
    pet: Pet
    n: int

print(Model.model_json_schema()['properties']['pet']['discriminator'])

"""
>>> {'mapping': {'cat': {'discriminator': {'mapping': {'black': '#/$defs/BlackCat', 'white': '#/$defs/WhiteCat'}, 'propertyName': 'color'}, 'oneOf': [{'$ref': '#/$defs/BlackCat'}, {'$ref': '#/$defs/WhiteCat'}]}, 'dog': '#/$defs/Dog'}, 'propertyName': 'pet_type'}

pydantic/main.py raises a validation_error when I ask for FastAPI to generate OpenAPI schema for these models:
pydantic.error_wrappers.ValidationError: 2 validation errors for OpenAPI
components -> schemas -> Model -> properties -> pet -> discriminator -> mapping -> cat
  str type expected (type=type_error.str)
components -> schemas -> Model -> $ref
  field required (type=value_error.missing)
"""
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0.3
pydantic-core version: 2.3.0 release build profile
install path: /Users/luke/Documents/repos/saas/webservice/.venv/lib/python3.10/site-packages/pydantic
python version: 3.10.10 (main, Jul 21 2023, 10:17:24) [Clang 14.0.0 (clang-1400.0.29.202)]
platform: macOS-12.5-x86_64-i386-64bit
optional deps. installed: ['email-validator', 'typing-extensions']
```

**Comment by samuelcolvin** [comment]
Could it be related to #7487?

**Comment by LukeHartmanTmlt** [comment]
> Could it be related to #7487?

It might be, but my current understanding is that #7487 is more of a typical bug where Pydantic is just doing the wrong thing, where this issue is more about Pydantic not being aligned with an external spec (OpenAPI) that, afaik, Pydantic is trying to match.

**Comment by LukeHartmanTmlt** [comment]
For example, this works correctly:
```
Model.model_validate({
    'pet': {
        'pet_type': 'cat',
        'color': 'black',
        'black_name': 'shadow',
    },
    'n': 1,
})
```

**Comment by CallumMcMahon** [comment]
I'm able to get a working FastAPI/OpenAPI schema from nested discriminated Unions using the solution mentioned in #7487
I only needed the TypeAliasType as I'm not using recursive models.
Mentioning in case this helps unblock anyone else short term!

**Comment by verngutz** [comment]
Can you share how exactly the solution mentioned in #7487 can be applied here? I tried all combinations of changing the parent union or the child unions into TypeAliasType to no avail.

**Comment by CallumMcMahon** [comment]
Continuing the example code from above
```python
from typing_extensions import Annotated, TypeAliasType

Cat = TypeAliasType(
    "Cat", 
    Annotated[Union[BlackCat, WhiteCat], Field(discriminator='color')]
)
print(str(Model.model_json_schema()["properties"]["pet"]))
print(str(Model.model_json_schema()["$defs"]["Cat"]))
```
This returns
```json
{
    "discriminator": {
        "mapping": {
            "black": "#/$defs/BlackCat",
            "white": "#/$defs/WhiteCat"
        },
        "propertyName": "color"
    },
    "oneOf": [
        {"$ref": "#/$defs/BlackCat"},
        {"$ref": "#/$defs/WhiteCat"}
    ]
}
{
    "discriminator": {
        "mapping": {
            "cat": "#/$defs/Cat",
            "dog": "#/$defs/Dog"
        },
        "propertyName": "pet_type"
    },
    "oneOf": [
        {"$ref": "#/$defs/Cat"},
        {"$ref": "#/$defs/Dog"}
    ],
    "title": "Pet"
}
```
Then confirmed with the following FastAPI code, which errors out prior to wrapping the type in a TypeAliasType
```python
from fastapi import APIRouter, FastAPI

app = FastAPI()
router = APIRouter()

@router.get("/")
def demo(input: Model):
    pass

app.include_router(router)
print(app.openapi())
```
@verngutz Hopefully this helps? Though I likely won't be looking at this comment thread again until after the festive period...

**Comment by verngutz** [comment]
@CallumMcMahon Thank you. Thanks to this, I found out that my real issue was using the `TypeAliasType` `Cat` to annotate another model's field, in addition to using the top-level discriminated union `Pet`. I fixed it by having both a normal union `Cat` and a `CatT = TypeAliasType("CatT", Cat)`, using `Cat` to annotate the field, and using `CatT` in the definition of `Pet`.

**Comment by sydney-runkle** [comment]
@LukeHartmanTmlt,

What schema would you expect?

**Comment by MickVermeulen** [comment]
@sydney-runkle I think the following FastAPI issue goes into detail more clearly: https://github.com/tiangolo/fastapi/discussions/11184. Do you think the generated schema here is correct, in the issue is instead on FastAPI on not supporting this type of schema correctly?

**Comment by sydney-runkle** [comment]
@MickVermeulen,

Thanks for the feedback - could you open a separate issue with your MRE and schema issues? Happy to help!

**Comment by AmirHmZz** [comment]
Any updates on fixing this?

**Comment by sydney-runkle** [comment]
Not at the moment - PRs welcome from the community in the meantime!

**Comment by 1313e** [comment]
Are there any updates? I now am also running into this issue. Can of course remove the outer discriminator, but that adds a performance penalty that nested discriminated unions is designed to prevent.

**Comment by Skitionek** [comment]
Wanted to give it a try but honestly cannot find the time:
https://github.com/Skitionek/fastapi/pull/1
anybody would like to test if this fits the need or point directions on what's needed instead?
